File: 8vim/src/main/java/inc/flide/vim8/structures/LayoutFileName.java
package inc.flide.vim8.structures;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import inc.flide.vim8.keyboardhelpers.KeyboardDataYamlParser;
import java.io.InputStream;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import org.apache.commons.lang3.StringUtils;

public class LayoutFileName {
    private static final Set<String> ISO_LANGUAGES = new HashSet<>(Arrays.asList(Locale.getISOLanguages()));
    private String languageCode;
    private String languageName;
    private boolean isValidLayout;
    private String resourceName;
    private String layoutDisplayName;

    private int totalLayers;

    public LayoutFileName() {
        languageCode = "en";
        languageName = "English";
        isValidLayout = true;
        resourceName = languageCode;
        layoutDisplayName = StringUtils.capitalize(languageName);
        totalLayers = 1;
    }

    @SuppressLint("DiscouragedApi")
    public LayoutFileName(Resources resources, Context context, String fileName) {
        this();
        if (ISO_LANGUAGES.contains(fileName)) {
            resourceName = fileName;
            languageCode = fileName;
            languageName = Locale.forLanguageTag(languageCode).getDisplayName(new Locale(languageCode));
            layoutDisplayName = StringUtils.capitalize(languageName);
            @SuppressLint("DiscouragedApi") int resourceId = resources.getIdentifier(fileName, "raw", context.getPackageName());
            try (InputStream inputStream = resources.openRawResource(resourceId)) {
                totalLayers = KeyboardDataYamlParser.isValidFile(inputStream);
                isValidLayout = true;
                if (totalLayers > 1) {
                    layoutDisplayName += " (" + totalLayers + " layers)";
                }
                if (totalLayers == 0) {
                    setLayoutValidityFalse();
                }
            } catch (Exception e) {
                setLayoutValidityFalse();
            }
        } else {
            setLayoutValidityFalse();
        }
    }

    private void setLayoutValidityFalse() {
        this.setValidLayout(false);
        this.languageCode = "";
        this.languageName = "";
        this.totalLayers = 0;
    }

    public String getResourceName() {
        return resourceName;
    }

    public String getLayoutDisplayName() {
        return layoutDisplayName;
    }

    public boolean isValidLayout() {
        return isValidLayout;
    }

    public void setValidLayout(boolean validLayout) {
        isValidLayout = validLayout;
    }
}

File: 8vim/src/main/java/inc/flide/vim8/structures/LayoutFileName.kt
package inc.flide.vim8.structures

import org.apache.commons.lang3.StringUtils
import java.util.*

class LayoutFileName() {
    private var languageCode: String = "en"
    private var fontCode: String = "regular"
    private var layoutName: String = "8pen"
    private var languageName: String = "English"
    private var isValidLayout = true
    private var resourceName: String
    private var layoutDisplayName: String?

    fun getResourceName() : String {
        return resourceName
    }

    fun isValidLayout() : Boolean {
        return isValidLayout
    }

    fun getLayoutDisplayName() : String? {
        return layoutDisplayName
    }

    constructor(fileName: String) : this() {
        val nameComponents: Array<String?> = fileName.split("_", ignoreCase = false, limit = 3).toTypedArray()
        if (nameComponents.size != 3) {
            setLayoutValidityFalse()
            return
        }
        if (ISO_LANGUAGES.contains(nameComponents[0]) && FONT_CODES.contains(nameComponents[1])) {
            resourceName = fileName
            languageCode = nameComponents[0]!!
            fontCode = nameComponents[1]!!
            layoutName = nameComponents[2]!!
            languageName = Locale.forLanguageTag(languageCode).getDisplayName(Locale(languageCode))
            layoutDisplayName = StringUtils.capitalize(layoutName) + " (" + StringUtils.capitalize(languageName) + ")"
            isValidLayout = true
        } else {
            setLayoutValidityFalse()
        }
    }

    private fun setLayoutValidityFalse() {
        isValidLayout = false
        languageCode = ""
        fontCode = ""
        layoutName = ""
        languageName = ""
    }

    companion object {
        private val ISO_LANGUAGES: MutableSet<String?> = HashSet(listOf(*Locale.getISOLanguages()))
        private val FONT_CODES: MutableSet<String?> = HashSet(
                listOf(
                "regular",
                "bold",
                "italic",
                "underline"
            )
        )
    }

    init {
        resourceName = languageCode + "_" + fontCode + "_" + layoutName
        layoutDisplayName = StringUtils.capitalize(layoutName) + " (" + StringUtils.capitalize(languageName) + ")"
    }
}